## Quick Analysis of Boston Housing Data

I will conduct an initial, quick analysis of the Boston Housing Data provided by sklearn in order to run some linear regressions using statsmodel. I will be building on the guide from Towards Data Science https://towardsdatascience.com/simple-and-multiple-linear-regression-in-python-c928425168f9.

This analysis is by no means thorough and I will use this as a starting point for a deeper analysis in the future.

For now, let's focus on testing out some regressions.


### Step 1: Import the data and packages

In [18]:
import statsmodels.api as sm
import numpy as np
import pandas as pd

In [19]:
# Import datasets from sklearn
from sklearn import datasets

# Read in the data from the datasets library
data = datasets.load_boston()

### Step 2: Begin to get familiar with the dataset

Note: Scikit-learn has already set the house value/price data as a target variable and 13 other variables are set as predictors.

In [20]:
# Show the column names (independent variables)
data.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], 
      dtype='<U7')

#### Description of the Boston House Prices dataset

##### Notes
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
        - LSTAT    % lower status of the population
        - MEDV     Median value of owner-occupied homes in $1000's

    :Missing Attribute Values: None

    :Creator: Harrison, D. and Rubinfeld, D.L.

This is a copy of UCI ML housing dataset.
http://archive.ics.uci.edu/ml/datasets/Housing


This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University.

In [21]:
# Show the dependent variable
data.target

array([ 24. ,  21.6,  34.7,  33.4,  36.2,  28.7,  22.9,  27.1,  16.5,
        18.9,  15. ,  18.9,  21.7,  20.4,  18.2,  19.9,  23.1,  17.5,
        20.2,  18.2,  13.6,  19.6,  15.2,  14.5,  15.6,  13.9,  16.6,
        14.8,  18.4,  21. ,  12.7,  14.5,  13.2,  13.1,  13.5,  18.9,
        20. ,  21. ,  24.7,  30.8,  34.9,  26.6,  25.3,  24.7,  21.2,
        19.3,  20. ,  16.6,  14.4,  19.4,  19.7,  20.5,  25. ,  23.4,
        18.9,  35.4,  24.7,  31.6,  23.3,  19.6,  18.7,  16. ,  22.2,
        25. ,  33. ,  23.5,  19.4,  22. ,  17.4,  20.9,  24.2,  21.7,
        22.8,  23.4,  24.1,  21.4,  20. ,  20.8,  21.2,  20.3,  28. ,
        23.9,  24.8,  22.9,  23.9,  26.6,  22.5,  22.2,  23.6,  28.7,
        22.6,  22. ,  22.9,  25. ,  20.6,  28.4,  21.4,  38.7,  43.8,
        33.2,  27.5,  26.5,  18.6,  19.3,  20.1,  19.5,  19.5,  20.4,
        19.8,  19.4,  21.7,  22.8,  18.8,  18.7,  18.5,  18.3,  21.2,
        19.2,  20.4,  19.3,  22. ,  20.3,  20.5,  17.3,  18.8,  21.4,
        15.7,  16.2,

### Step 3: Run a linear regression on this dataset

#### 3A. Set the predictors (independent variables) and the target (dependent variable we're trying to predict)

In [22]:
# Load the data as a pandas data frame for easier analysis 
# Define the data/predictors as the pre-set feature names  
df = pd.DataFrame(data.data, columns=data.feature_names)

# Preview the first five rows of the dataframe
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [23]:
# Set the median home value as our target variable
# Put the target (housing value: MEDV) in another dataframe
target = pd.DataFrame(data.target, columns=["MEDV"])

# Preview the first five rows of the dataframe
target.head()

MEDV
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

#### 3B. Fit a linear regression model

Choose variables that we think will be good predictors of the target

I will start with RM (the average number of rooms) and fit the regression model with one variable

In [34]:
# Independent variable (Input)
rm = df["RM"]
rm = sm.add_constant(rm) ## Add an intercept (beta_0) to our model

# Dependent variable (Output)
medv = target["MEDV"]

# Create the model (ordinary least squares-OLS)
model = sm.OLS(medv, rm).fit()

# Make predictions from the model
predictions = model.predict(rm)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.484
Model:                            OLS   Adj. R-squared:                  0.483
Method:                 Least Squares   F-statistic:                     471.8
Date:                Wed, 07 Mar 2018   Prob (F-statistic):           2.49e-74
Time:                        14:00:22   Log-Likelihood:                -1673.1
No. Observations:                 506   AIC:                             3350.
Df Residuals:                     504   BIC:                             3359.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -34.6706      2.650    -13.084      0.000     -39.877     -29.465
RM             9.1021      0.419     21.722      0.000       8.279       9.925
==============================================================================
Omnibus:                      102.585   Durbin-Watson:                   0.684
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              612.449
Skew:                           0.726   Prob(JB):                    1.02e-133
Kurtosis:                       8.190   Cond. No.                         58.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The r-squared (0.484) shows that number of rooms (RM) describes 48.4% of the variance in median home values. 

The constant (y-intercept) is -34.6706. Without a constant, we'd be forcing the model to go through the origin.

The slope of the RM predictor is 9.1021.

#### Now, try fitting a regression model with multiple variables, RM and LSTAT (percentage of lower status of the population).

In [35]:
# Independent variables (Input)
X = df[["RM", "LSTAT"]]

# Dependent variable (Output)
y = target["MEDV"]

# Create the model (ordinary least squares-OLS)
model = sm.OLS(y, X).fit()

# Make predictions from the model
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.948
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     4637.
Date:                Wed, 07 Mar 2018   Prob (F-statistic):               0.00
Time:                        14:01:35   Log-Likelihood:                -1582.9
No. Observations:                 506   AIC:                             3170.
Df Residuals:                     504   BIC:                             3178.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
RM             4.9069      0.070     69.906      0.000       4.769       5.045
LSTAT         -0.6557      0.031    -21.458      0.000      -0.716      -0.596
==============================================================================
Omnibus:                      145.153   Durbin-Watson:                   0.834
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              442.157
Skew:                           1.351   Prob(JB):                     9.70e-97
Kurtosis:                       6.698   Cond. No.                         4.72
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The r-squared (0.948) shows that these two variables describe 94.8% of the variance in median home values. That's already pretty good!

#### Now, I will extend the example and try fitting a regression model with three variables, RM, LSTAT, CRIM (per capita crime rate by town).

In [37]:
# Independent variables (Input)
X = df[["RM", "LSTAT", "CRIM"]]

# Dependent variable (Output)
y = target["MEDV"]

# Create the model (ordinary least squares-OLS)
model = sm.OLS(y, X).fit()

# Make predictions from the model
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.949
Model:                            OLS   Adj. R-squared:                  0.949
Method:                 Least Squares   F-statistic:                     3147.
Date:                Wed, 07 Mar 2018   Prob (F-statistic):               0.00
Time:                        14:01:53   Log-Likelihood:                -1578.1
No. Observations:                 506   AIC:                             3162.
Df Residuals:                     503   BIC:                             3175.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
RM             4.8649      0.071     68.602      0.000       4.726       5.004
LSTAT         -0.6065      0.034    -17.714      0.000      -0.674      -0.539
CRIM          -0.0982      0.032     -3.093      0.002      -0.161      -0.036
==============================================================================
Omnibus:                      168.770   Durbin-Watson:                   0.820
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              579.077
Skew:                           1.533   Prob(JB):                    1.80e-126
Kurtosis:                       7.250   Cond. No.                         5.20
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-squared didn't go up very much by adding CRIM. It's .001 higher (0.949).

What if we remove CRIM and add DIS (weighted distances to five Boston employment centres)?

In [38]:
# Independent variables (Input)
X = df[["RM", "LSTAT", "DIS"]]

# Dependent variable (Output)
y = target["MEDV"]

# Create the model (ordinary least squares-OLS)
model = sm.OLS(y, X).fit()

# Make predictions from the model
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.950
Model:                            OLS   Adj. R-squared:                  0.949
Method:                 Least Squares   F-statistic:                     3157.
Date:                Wed, 07 Mar 2018   Prob (F-statistic):               0.00
Time:                        14:03:16   Log-Likelihood:                -1577.3
No. Observations:                 506   AIC:                             3161.
Df Residuals:                     503   BIC:                             3173.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
RM             5.2305      0.119     43.902      0.000       4.996       5.465
LSTAT         -0.6921      0.032    -21.526      0.000      -0.755      -0.629
DIS           -0.4206      0.126     -3.344      0.001      -0.668      -0.173
==============================================================================
Omnibus:                      120.746   Durbin-Watson:                   0.884
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              352.275
Skew:                           1.133   Prob(JB):                     3.19e-77
Kurtosis:                       6.402   Cond. No.                         10.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-squared went up ever so slightly more, but barely. It's only .002 higher than our two-variable regression. (0.950)